design-proposals/volume-migration/volume-migration.md

```diff
@@ -0,0 +1,315 @@
+
+## Repos
+
+  - [kubevirt/kubevirt](https://github.com/kubevirt/kubevirt)
+
+# Design
+This proposal introduces a new CRD called `VolumeMigration`
+
+```golang
+type VolumeMigration struct {
+	metav1.TypeMeta   `json:",inline"`
+	metav1.ObjectMeta `json:"metadata,omitempty"`
+	Spec VolumeMigrationSpec `json:"spec"`
+	Status *VolumeMigrationStatus `json:"status,omitempty"`
+}
+
+type VolumeMigrationList struct {
+	metav1.TypeMeta `json:",inline"`
+	metav1.ListMeta `json:"metadata"`
+	Items []VolumeMigration `json:"items"`
+}
+
+type ReclaimPolicySourcePvc string
+
+const (
+	DeleteReclaimPolicySourcePvc ReclaimPolicySourcePvc = "Delete"
+	RetainReclaimPolicySourcePvc ReclaimPolicySourcePvc = "Retain"
+)
+
+type MigratedVolume struct {
+	SourcePvc      string `json:"sourcePvc,omitempty" valid:"required"`
+	DestinationPvc string `json:"destinationPvc,omitempty" valid:"required"`
+	ReclaimPolicySourcePvc ReclaimPolicySourcePvc `json:"reclaimPolicySourcePvc,omitempty" valid:"required"`
+}
+
+type VolumeMigrationSpec struct {
+	MigratedVolume []MigratedVolume `json:"migratedVolume,omitempty"`
+}
+
+type RejectedMigratedVolume struct {
+	SourcePvc string `json:"sourcePvc,omitempty" valid:"required"`
+	Reason    string `json:"reason,omitempty" valid:"required"`
+}
+
+type RejectedVolumeMigration struct {
+	RejectedVolumes []RejectedMigratedVolume `json:"rejectedVolumes,omitempty" valid:"required"`
+	AssociatedVolumes []string `json:"associatedVolumes,omitempty"`
+	Reason            string   `json:"reason,omitempty" valid:"required"`
+}
+
+type VolumeMigrationPhase string
+
+const (
+	VolumeMigrationPhaseFailed     = "Failed"
+	VolumeMigrationPhasePending    = "Pending"
+	VolumeMigrationPhaseRejected   = "Rejected"
+	VolumeMigrationPhaseRunning    = "Running"
+	VolumeMigrationPhaseScheduling = "Scheduling"
+	VolumeMigrationPhaseSucceeded  = "Succeeded"
+	VolumeMigrationPhaseUnknown    = "Unknown"
+)
+
+type VolumeMigrationPhaseTransitionTimestamp struct {
+	Phase                    VolumeMigrationPhase `json:"phase,omitempty"`
+	PhaseTransitionTimestamp metav1.Time          `json:"phaseTransitionTimestamp,omitempty"`
+}
+
+type VolumeMigrationState struct {
+	MigratedVolume []MigratedVolume `json:"migratedVolume,omitempty"`
+	VirtualMachineMigrationName string `json:"virtualMachineMigrationName,omitempty"`
+	Succeeded bool `json:"succeeded,omitempty"`
+	Failed bool `json:"failed,omitempty"`
+	StartTimestamp *metav1.Time `json:"startTimestamp,omitempty"`
+	EndTimestamp *metav1.Time `json:"endTimestamp,omitempty"`
+}
+
+type VolumeMigrationStatus struct {
+	VirtualMachineInstanceName string                  `json:"virtualMachineInstanceName,omitempty"`
+	VolumeMigrationState       VolumeMigrationState    `json:"volumeMigrationState,omitempty"`
+	PendingVolumes             []MigratedVolume        `json:"pendingVolumes,omitempty"`
+	RejectedVolumeMigration    RejectedVolumeMigration `json:"rejectedVolumeMigration,omitempty"`
```

Comment:
I did in this way because `VolumeMigrationState` has the infomation for the migration while in the case on pending and rejected volumes the VM migration isn't triggered.

About the VMI name is because rejected volumes are associated to a VMI but not to a VMI migration. 
However, we can also leave them empty and have everything inside the same struct.

We could also simplify futher the status and remove the Succeeded and Failed bools.
Like:
```golang
type VolumeMigrationState struct {
	VirtualMachineInstanceName  string `json:"virtualMachineInstanceName,omitempty"`
	VirtualMachineMigrationName string `json:"virtualMachineMigrationName,omitempty"`

	MigratedVolumes         []MigratedVolume         `json:"migratedVolume,omitempty"`
	PendingVolumes          []MigratedVolume         `json:"pendingVolumes,omitempty"`
	RejectedVolumeMigration *RejectedVolumeMigration `json:"rejectedVolumeMigration,omitempty"`

	StartTimestamp *metav1.Time `json:"startTimestamp,omitempty"`
	EndTimestamp   *metav1.Time `json:"endTimestamp,omitempty"`
}

type VolumeMigrationStatus struct {
	VolumeMigrationState *VolumeMigrationState `json:"volumeMigrationState,omitempty"`
	Phase                VolumeMigrationPhase  `json:"phase,omitempty"`
	// PhaseTransitionTimestamp is the timestamp of when the last phase change occurred
	PhaseTransitionTimestamps []VolumeMigrationPhaseTransitionTimestamp `json:"phaseTransitionTimestamps,omitempty"`
}
```

Reply:
Can we go even further and simplify:

```go
	MigratedVolumes         []MigratedVolume         `json:"migratedVolume,omitempty"`
	PendingVolumes          []MigratedVolume         `json:"pendingVolumes,omitempty"`
	RejectedVolumeMigration *RejectedVolumeMigration `json:"rejectedVolumeMigration,omitempty"`
```

Probably we could introduce `type VolumeMigrationState struct` which combines the fields and then have a flat list:

```go
	VolumeMigrationStates         []VolumeMigrationState         `json:"volumeMigrationStates,omitempty"`
```

WDYT?

Reply:
I'm not sure I understand your suggestion.

Do you mean something like this:
```golang
type VolumeMigrationStatus struct {
        VirtualMachineInstanceName  string                                    `json:"virtualMachineInstanceName,omitempty"`
        VirtualMachineMigrationName string                                    `json:"virtualMachineMigrationName,omitempty"`
        StartTimestamp              *metav1.Time                              `json:"startTimestamp,omitempty"`
        EndTimestamp                *metav1.Time                              `json:"endTimestamp,omitempty"`
        ValidVolumes                []MigratedVolume                          `json:"validVolumes,omitempty"`
        PendingVolumes              []MigratedVolume                          `json:"pendingVolumes,omitempty"`
        RejectedVolumes             []MigratedVolume                          `json:"rejectedVolumeMigration,omitempty"`
        Conditions                  []Conditions                              `json:"conditions,omitempty"`
        Phase                       VolumeMigrationPhase                      `json:"phase,omitempty"`
        PhaseTransitionTimestamps   []VolumeMigrationPhaseTransitionTimestamp `json:"phaseTransitionTimestamps,omitempty"`
}
```

Reply:
Or something like this:
```golang 
type MigratedVolumeState string

const (
        MigratedVolumeStateValid    = "Valid"
        MigratedVolumeStatePending  = "Pending"
        MigratedVolumeStateRejected = "Rejected"
)

type VolumeMigrationState struct {
        SourcePvc string              `json:"sourcePvc,omitempty"`
        State     MigratedVolumeState `json:"state,omitempty"`
        Reason    string              `json:"reason,omitempty"`
}

type VolumeMigrationStatus struct {
        VolumeMigrationStates       []VolumeMigrationState                    `json:"volumeMigrationStates,omitempty"`
        VirtualMachineInstanceName  string                                    `json:"virtualMachineInstanceName,omitempty"`
        VirtualMachineMigrationName string                                    `json:"virtualMachineMigrationName,omitempty"`
        StartTimestamp              *metav1.Time                              `json:"startTimestamp,omitempty"`
        EndTimestamp                *metav1.Time                              `json:"endTimestamp,omitempty"`
        Phase                       VolumeMigrationPhase                      `json:"phase,omitempty"`
        PhaseTransitionTimestamps   []VolumeMigrationPhaseTransitionTimestamp `json:"phaseTransitionTimestamps,omitempty"`
}

```

Reply:
Ah, sorry. I probably expressed it badly. I was thinking if it could make sense to merge
```go
	MigratedVolumes         []MigratedVolume         `json:"migratedVolume,omitempty"`
	PendingVolumes          []MigratedVolume         `json:"pendingVolumes,omitempty"`
	RejectedVolumeMigration *RejectedVolumeMigration `json:"rejectedVolumeMigration,omitempty"`
```
into one list.

Smth like in the code below:

```go
// struct with the current status for a single volume
type VolumeStatus struct {
	MigratedVolume MigratedVolume
	Status string //pending, rejected, completed, etc.
	Reason string
	... //maybe other fields
}

type VolumeMigrationState struct {
	VirtualMachineInstanceName  string `json:"virtualMachineInstanceName,omitempty"`
	VirtualMachineMigrationName string `json:"virtualMachineMigrationName,omitempty"`

	// one list instead of	
	//MigratedVolumes         []MigratedVolume         `json:"migratedVolume,omitempty"`
	//PendingVolumes          []MigratedVolume         `json:"pendingVolumes,omitempty"`
	//RejectedVolumeMigration *RejectedVolumeMigration `json:"rejectedVolumeMigration,omitempty"`
	VolumeStatuses []VolumeStatuses `json:"volumeStatuses,omitempty"`

	StartTimestamp *metav1.Time `json:"startTimestamp,omitempty"`
	EndTimestamp   *metav1.Time `json:"endTimestamp,omitempty"`
}

type VolumeMigrationStatus struct {
	VolumeMigrationState *VolumeMigrationState `json:"volumeMigrationState,omitempty"`
	Phase                VolumeMigrationPhase  `json:"phase,omitempty"`
	// PhaseTransitionTimestamp is the timestamp of when the last phase change occurred
	PhaseTransitionTimestamps []VolumeMigrationPhaseTransitionTimestamp `json:"phaseTransitionTimestamps,omitempty"`
}
```

Reply:
ok :) make sense to me, thanks!

Reply:
Updated with the suggested changes